src/Kestrel.Core/Internal/Http/MessageBody.cs

```diff
@@ -43,17 +48,27 @@ public virtual async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationT
 
             while (true)
             {
-                var result = await _context.RequestBodyPipe.Reader.ReadAsync(cancellationToken);
+                var result = await StartTimingReadAsync(cancellationToken);
                 var readableBuffer = result.Buffer;
+                var readableBufferLength = readableBuffer.Length;
+                StopTimingRead(readableBufferLength);
+
                 var consumed = readableBuffer.End;
                 var actual = 0;
 
                 try
                 {
                     if (!readableBuffer.IsEmpty)
                     {
-                        // buffer.Count is int
-                        actual = (int)Math.Min(readableBuffer.Length, buffer.Length);
+                        // buffer.Length is int
+                        actual = (int)Math.Min(readableBufferLength, buffer.Length);
+
+                        // Make sure we don't double-count bytes on the next read.
+                        if (readableBufferLength > actual)
+                        {
+                            _alreadyTimedBytes = readableBufferLength - actual;
```

Comment:
Can you get rid of _alreadyTimedBytes by calculating actual before StopTimingRead?

Reply:
We could, but I think it's to give credit to the client for sending data as soon as possible. Otherwise, if the app uses really small buffers and processes the data slowly once it arrives, we could unfairly penalize high-latency connections.